File: indra-common/src/main/java/net/kyori/indra/IndraExtension.java
/*
 * This file is part of indra, licensed under the MIT License.
 *
 * Copyright (c) 2020-2021 KyoriPowered
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package net.kyori.indra;

import net.kyori.indra.api.model.ApplyTo;
import net.kyori.indra.api.model.ContinuousIntegration;
import net.kyori.indra.api.model.Issues;
import net.kyori.indra.api.model.License;
import net.kyori.indra.api.model.SourceCodeManagement;
import net.kyori.mammoth.Configurable;
import org.gradle.api.Action;
import org.gradle.api.provider.Property;
import org.gradle.api.publish.maven.MavenPublication;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import static java.util.Objects.requireNonNull;

/**
 * Extension exposing extra functionality provided by Indra.
 *
 * @since 2.0.0
 */
public interface IndraExtension {
  /**
   * Options controlling Java toolchain versions.
   *
   * @since 2.0.0
   */
  @NotNull JavaToolchainVersions javaVersions();

  /**
   * Configure the versioning configuration.
   *
   * @since 2.0.0
   */
  void javaVersions(final @NotNull Action<JavaToolchainVersions> action);

  @NotNull Property<ContinuousIntegration> ci();

  default void ci(final @NotNull ContinuousIntegration ci) {
    this.ci().set(ci);
  }

  default void ci(final @NotNull Action<ContinuousIntegration.Builder> configureAction) {
    this.ci().set(Configurable.configure(ContinuousIntegration.builder(), configureAction).build());
  }

  default void jenkins(final @NotNull String url) {
    requireNonNull(url, "url");

    this.ci(ContinuousIntegration.builder()
      .system("Jenkins")
      .url(url)
      .build());
  }

  @NotNull Property<Issues> issues();

  default void issues(final @NotNull Issues issues) {
    this.issues().set(issues);
  }

  default void issues(final @NotNull Action<Issues.Builder> configureAction) {
    this.issues().set(Configurable.configure(Issues.builder(), configureAction).build());
  }

  @NotNull Property<SourceCodeManagement> scm();

  default void scm(final @NotNull SourceCodeManagement scm) {
    this.scm().set(scm);
  }

  default void scm(final @NotNull Action<SourceCodeManagement.Builder> configureAction) {
    this.scm().set(Configurable.configure(SourceCodeManagement.builder(), configureAction).build());
  }

  @NotNull Property<License> license();

  default void license(final @NotNull License license) {
    this.license().set(license);
  }

  default void license(final @NotNull Action<License.Builder> configureAction) {
    this.license().set(Configurable.configure(License.builder(), configureAction).build());
  }

  default void apache2License() {
    this.license(License.apache2());
  }

  default void gpl3OnlyLicense() {
    this.license(License.gpl3Only());
  }

  default void gpl3OrLaterLicense() {
    this.license(License.gpl3OrLater());
  }

  default void lgpl3OnlyLicense() {
    this.license(License.lgpl3Only());
  }

  default void lgpl3OrLaterLicense() {
    this.license(License.lgpl3OrLater());
  }

  default void mitLicense() {
    this.license(License.mit());
  }

  default void mpl2License() {
    this.license(License.mpl2());
  }

  default void github(final @NotNull String user, final @NotNull String repo) {
    this.github(user, repo, null);
  }

  void github(final @NotNull String user, final @NotNull String repo, final @Nullable Action<ApplyTo> applicable);

  default void gitlab(final @NotNull String user, final @NotNull String repo) {
    this.gitlab(user, repo, null);
  }

  void gitlab(final @NotNull String user, final @NotNull String repo, final @Nullable Action<ApplyTo> applicable);

  // Publishing repositories

  void publishAllTo(final @NotNull String id, final @NotNull String url);

  void publishReleasesTo(final @NotNull String id, final @NotNull String url);

  void publishSnapshotsTo(final @NotNull String id, final @NotNull String url);

  void configurePublications(final @NotNull Action<MavenPublication> action);

  /**
   * A property representing the version of checkstyle to be used.
   *
   * <p>If any custom additions are applied to the {@code checkstyle} configuration,
   * this value <b>will be ignored.</b></p>
   *
   * @return the checkstyle version property
   * @since 2.0.0
   */
  @NotNull Property<String> checkstyle();

  /**
   * Set the version of checkstyle to be used.
   *
   * @param checkstyleVersion the target checkstyle version
   * @see #checkstyle() for information on limitations
   * @since 2.0.0
   */
  default void checkstyle(final @NotNull String checkstyleVersion) {
    this.checkstyle().set(checkstyleVersion);
  }

  /**
   * Whether options that support reproducible builds should be enabled.
   *
   * <p>Default: {@code true}</p>
   *
   * @return the property configuring reproducible builds
   * @since 2.0.0
   */
  @NotNull Property<Boolean> reproducibleBuilds();

  /**
   * Set whether options that support reproducible builds should be enabled.
   *
   * <p>Default: {@code true}</p>
   *
   * @param reproducibleBuilds whether to build in a way that encourages reproducibility
   * @since 2.0.0
   */
  default void reproducibleBuilds(final boolean reproducibleBuilds) {
    this.reproducibleBuilds().set(reproducibleBuilds);
  }

  /**
   * Whether the {@code java} {@link org.gradle.api.component.SoftwareComponent} should be
   * automatically included in publications.
   *
   * <p>This property does not usually need to be changed, unless working with Gradle plugins
   * that publish in a non-standard way.</p>
   *
   * @return the property representing this option
   * @since 2.0.0
   */
  @NotNull Property<Boolean> includeJavaSoftwareComponentInPublications();

  /**
   * Set whether the {@code java} component should be included in publications.
   *
   * @param include whether the component should be included
   * @see #includeJavaSoftwareComponentInPublications() for more details
   * @since 2.0.0
   */
  default void includeJavaSoftwareComponentInPublications(final boolean include) {
    this.includeJavaSoftwareComponentInPublications().set(include);
  }
}

File: indra-common/src/main/kotlin/net/kyori/indra/IndraExtension.kt
/*
 * This file is part of indra, licensed under the MIT License.
 *
 * Copyright (c) 2020 KyoriPowered
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package net.kyori.indra

import java.net.URI
import javax.inject.Inject
import net.kyori.indra.api.model.ContinuousIntegration
import net.kyori.indra.api.model.Issues
import net.kyori.indra.api.model.License
import net.kyori.indra.api.model.SourceCodeManagement
import net.kyori.indra.data.ApplyTo
import net.kyori.indra.data.ContinuousIntegrationImpl
import net.kyori.indra.data.IssuesImpl
import net.kyori.indra.data.LicenseImpl
import net.kyori.indra.data.SourceCodeManagementImpl
import org.gradle.api.Action
import org.gradle.api.GradleException
import org.gradle.api.component.SoftwareComponent
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.Property
import org.gradle.api.publish.maven.MavenPublication
import org.gradle.kotlin.dsl.domainObjectSet
import org.gradle.kotlin.dsl.invoke
import org.gradle.kotlin.dsl.property
import org.gradle.process.CommandLineArgumentProvider

open class IndraExtension @Inject constructor(objects: ObjectFactory) {
  val reproducibleBuilds: Property<Boolean> = objects.property(Boolean::class).convention(true)

  /**
   * Whether the `java` [SoftwareComponent] should be automatically included in publications.
   *
   * This property does not usually need to be changed, unless working with Gradle plugins that publish in a non-standard way.
   */
  val includeJavaSoftwareComponentInPublications: Property<Boolean> = objects.property(Boolean::class).convention(true)

  /**
   * Options controlling Java toolchain versions
   */
  val javaVersions: JavaToolchainVersions = objects.newInstance(JavaToolchainVersions::class.java)

  /**
   * Configure the versioning configuration.
   */
  fun javaVersions(action: Action<JavaToolchainVersions>) {
    action(javaVersions)
  }

  internal fun previewFeatureArgumentProvider(): CommandLineArgumentProvider = CommandLineArgumentProvider {
    javaVersions.enablePreviewFeatures.finalizeValue()
    if(javaVersions.enablePreviewFeatures.get()) {
      listOf("--enable-preview")
    } else {
      listOf()
    }
  }

  val ci: Property<ContinuousIntegration> = objects.property(ContinuousIntegration::class)
  val issues: Property<Issues> = objects.property(Issues::class)
  val license: Property<License> = objects.property(License::class)
  val scm: Property<SourceCodeManagement> = objects.property(SourceCodeManagement::class)

  @JvmOverloads
  fun github(user: String, repo: String, applicable: Action<ApplyTo>? = null) {
    val options = ApplyTo().also { applicable?.execute(it) }
    if(options.ci) {
      this.ci.set(ContinuousIntegrationImpl("GitHub Actions", "https://github.com/$user/$repo/actions"))
    }
    if(options.issues) {
      this.issues.set(IssuesImpl("GitHub", "https://github.com/$user/$repo/issues"))
    }
    if(options.scm) {
      this.scm.set(SourceCodeManagementImpl("scm:git:https://github.com/$user/$repo.git", "scm:git:ssh://git@github.com/$user/$repo.git", "https://github.com/$user/$repo"))
    }
    if(options.publishing) {
      this.publishReleasesTo("githubPackages", "https://maven.pkg.github.com/$user/$repo")
    }
  }

  @JvmOverloads
  fun gitlab(user: String, repo: String, applicable: Action<ApplyTo>? = null) {
    val options = ApplyTo().also { applicable?.execute(it) }
    if(options.ci) {
      this.ci.set(ContinuousIntegrationImpl("GitLab CI", "https://gitlab.com/$user/$repo/-/pipelines"))
    }
    if(options.issues) {
      this.issues.set(IssuesImpl("GitLab", "https://gitlab.com/$user/$repo/-/issues"))
    }
    if(options.scm) {
      this.scm.set(SourceCodeManagementImpl("scm:git:https://gitlab.com/$user/$repo.git", "scm:git:ssh://git@gitlab.com/$user/$repo.git", "https://gitlab.com/$user/$repo"))
    }
    if(options.publishing) {
      // TODO: needs project ID, which is separate from user/repo and uses HTTP header-based auth
      throw GradleException("Publishing cannot yet be automatically configured for GitLab projects")
    }
  }

  fun apache2License() = this.license.set(LicenseImpl(
    spdx = "Apache-2.0",
    name = "Apache License, Version 2.0",
    url = "https://opensource.org/licenses/Apache-2.0"
  ))

  fun gpl3OnlyLicense() = this.license.set(LicenseImpl(
    spdx = "GPL-3.0-only",
    name = "GNU General Public License version 3",
    url = "https://opensource.org/licenses/GPL-3.0",
    bintray = "GPL-3.0"
  ))

  fun mitLicense() = this.license.set(LicenseImpl(
    spdx = "MIT",
    name = "The MIT License",
    url = "https://opensource.org/licenses/MIT"
  ))

  fun jenkins(url: String) = this.ci.set(ContinuousIntegrationImpl(
    system = "Jenkins",
    url = url
  ))

  // Checkstyle

  val checkstyle: Property<String> = objects.property(String::class).convention("8.37")

  // Publishing

  internal val repositories = objects.domainObjectSet(RepositorySpec::class)

  @Transient
  internal val publishingActions = mutableSetOf<Action<MavenPublication>>()

  fun publishAllTo(id: String, url: String) = this.repositories.add(RepositorySpec(id, URI(url), releases = true, snapshots = true))
  fun publishReleasesTo(id: String, url: String) = this.repositories.add(RepositorySpec(id, URI(url), releases = true, snapshots = false))
  fun publishSnapshotsTo(id: String, url: String) = this.repositories.add(RepositorySpec(id, URI(url), releases = false, snapshots = true))

  fun configurePublications(action: Action<MavenPublication>) {
    this.publishingActions.add(action)
  }
}

internal data class RepositorySpec(
  val id: String,
  val url: URI,
  val releases: Boolean,
  val snapshots: Boolean
)
